pxtblocks/fields/field_melodySandbox.ts

```diff
@@ -0,0 +1,415 @@
+        }
+
+        protected getDropdownBorderColour() {
+            return this.sourceBlock_.getColourSecondary();
+            //return "#560649";
+        }
+
+        setTitle(melodyName: string): void {
+            if (this.title != melodyName) {
+                this.title = melodyName;
+                Blockly.FieldLabel.prototype.setText.call(this, this.title);
+                if (this.melody) {
+                    this.melody.setTitle(this.title);
+                }
+            }
+        }
+
+        getTitle(): string {
+            return this.title;
+        }
+
+        setTempo(tempo: number): void {
+            // reset text input if input is invalid
+            if (isNaN(tempo) || tempo <= 0 && document.getElementById("melody-tempo-text")) {
+                (<HTMLInputElement>document.getElementById("melody-tempo-text")).value = this.tempo + "";
+                return
+            }
+            // update tempo and duration values and display to reflect new tempo
+            if (this.tempo != tempo) {
+                this.tempo = tempo;
+                this.duration = 60000 / tempo;
+                if (this.melody) {
+                    this.melody.setTempo(this.tempo);
+                }
+                if (document.getElementById("melody-tempo-text")) {
+                    (<HTMLInputElement>document.getElementById("melody-tempo-text")).value = this.tempo + "";
+                }
+                if (document.getElementById("melody-tempo-slider")) {
+                    (<HTMLInputElement>document.getElementById("melody-tempo-slider")).value = this.tempo + "";
+                }
+            }
+        }
+
+        onNoteSelect(id: string): void {
+            // parse element id
+            let params = id.split("-"); // params[1] is row, params[2] is cell
+            let row = params[1];
+            let col = params[2];
+
+
+            // play sound if selected
+            if (!this.melody.getValue(+row, +col)) {
+                this.playNote(row);
+                if (this.oneNotePerCol) { // clear all other notes in col
+                    for (let i = 0; i < this.numRow; i++) {
+                        if (this.melody.getValue(i, +col)) {
+                            // update melody array
+                            this.melody.updateMelody(i, +col);
+                            // set color to default
+                            document.getElementById("cell-" + i + "-" + col).style.backgroundColor = "";
+                        }
+                    }
+
+                }
+                // update button/div color
+                this.updateColor(id, +row, +col);
+            } else {
+                // set color to default
+                document.getElementById(id).style.backgroundColor = "";
+            }
+
+            // update melody array
+            this.melody.updateMelody(+row, +col);
+        }
+
+
+
+        playNote(rowNumber: string, colNumber?: number): void {
```

Comment:
It's a string because I parse a string to get the row number, but I can change it and just convert to a number before I call it. I was also thinking it might make more sense for to have an enum with all the note information instead of having the switch statements in playNote() and updateColor().

Reply:
Yeah, lets make it a number